### PR TITLE
rgw: Project Zipper - Bucketlist

### PR DESCRIPTION
**Pull request by dang**
Implement and do the initial conversion to the BucketList API for
Zipper.  This invloves doing portions of the User and Bucket API as
well, so enable BucketList to be used.

This is not the final version; the API needs to be pushed further down
into RGWRados code, but this gets the API layer complete.

Signed-off-by: Daniel Gryniewicz <dang@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -149,17 +149,15 @@ void rgw_parse_url_bucket(const string &bucket, const string& auth_tenant,
  */
 int rgw_read_user_buckets(RGWRadosStore * store,
                           const rgw_user& user_id,
-                          RGWUserBuckets& buckets,
+                          rgw::sal::RGWBucketList& buckets,
```
Comment (mattbenjamin) [inline src/rgw/rgw_bucket.cc]: would try for a different namespace name here--maybe "storage?"

File: src/rgw/rgw_bucket.cc [hunk 2]
```diff
@@ -255,48 +253,44 @@ static void dump_mulipart_index_results(list<rgw_obj_index_key>& objs_to_unlink,
...
+    int ret = user.list_buckets(marker, string(), max_entries, false, user_buckets);
     if (ret < 0) {
       ldout(store->ctx(), 0) << "failed to read user buckets: "
 			     << cpp_strerror(-ret) << dendl;
       return;
     }
 
-    map<string, RGWBucketEnt>& buckets = user_buckets.get_buckets();
-    for (map<string, RGWBucketEnt>::iterator i = buckets.begin();
+    map<string, RGWSalBucket*>& buckets = user_buckets.get_buckets();
```
Comment (mattbenjamin) [inline src/rgw/rgw_bucket.cc]: and rely on namespace qualification here, and related

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -2716,22 +2715,9 @@ void RGWStatBucket::execute()
...
-    op_ret = 0;
-    map<string, RGWBucketEnt>::iterator iter = m.find(bucket.bucket.name);
-    if (iter != m.end()) {
-      bucket = iter->second;
-    } else {
-      op_ret = -EINVAL;
-    }
-  }
+  rgw::sal::RGWRadosUser user(store, s->user->user_id);
+  bucket = new rgw::sal::RGWRadosBucket(store, user, s->bucket);
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: what's happening here--operating directly on RGWRadosBucket?

File: src/rgw/rgw_op.cc [hunk 4]
```diff
@@ -4784,7 +4763,7 @@ void RGWDeleteObj::execute()
     obj_ctx->set_atomic(obj);
 
     bool ver_restored = false;
-    op_ret = store->getRados()->swift_versioning_restore(*s->sysobj_ctx, *obj_ctx, s->bucket_owner.get_id(),
+    op_ret = store->getRados()->swift_versioning_restore(*obj_ctx, s->bucket_owner.get_id(),
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: nice cleanup (picking his instance at random)

File: src/rgw/rgw_rados.h [hunk 5]
```diff
@@ -1310,12 +1309,13 @@ class RGWRados
...
 
   int put_bucket_instance_info(RGWBucketInfo& info, bool exclusive, ceph::real_time mtime, map<string, bufferlist> *pattrs);
+  /* xxx dang obj_ctx -> svc */
   int get_bucket_instance_info(RGWSysObjectCtx& obj_ctx, const string& meta_key, RGWBucketInfo& info, ceph::real_time *pmtime, map<string, bufferlist> *pattrs, optional_yield y);
   int get_bucket_instance_info(RGWSysObjectCtx& obj_ctx, const rgw_bucket& bucket, RGWBucketInfo& info, ceph::real_time *pmtime, map<string, bufferlist> *pattrs, optional_yield y);
 
   static void make_bucket_entry_name(const string& tenant_name, const string& bucket_name, string& bucket_entry);
 
-  int get_bucket_info(RGWSysObjectCtx& obj_ctx,
+  int get_bucket_info(RGWServices *svc,
```
Comment (mattbenjamin) [inline src/rgw/rgw_rados.h]: and here

File: src/rgw/rgw_rest_s3.h [hunk 6]
```diff
@@ -113,7 +112,7 @@ class RGWListBuckets_ObjStore_S3 : public RGWListBuckets_ObjStore {
     return 0;
   }
   void send_response_begin(bool has_buckets) override;
-  void send_response_data(RGWUserBuckets& buckets) override;
+  void send_response_data(rgw::sal::RGWBucketList& buckets) override;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.h]: at the very least 
rgw::sal::Bucket is preferable to RGWSalBucket--but again, I think I'm liking rgw::storage::Bucket

File: src/rgw/rgw_bucket.cc (hunk 1 above)
Comment (dang) [reply src/rgw/rgw_bucket.cc]: I'm not sure I understand.  Are you objecting to sal?

File: src/rgw/rgw_bucket.cc (hunk 2 above)
Comment (dang) [reply src/rgw/rgw_bucket.cc]: RGWBucket exists, and is needed (currenly) by the SAL code.  The plan is to rename RGWSalBucket to RGWBucket when it's possible.

File: src/rgw/rgw_op.cc (hunk 3 above)
Comment (dang) [reply src/rgw/rgw_op.cc]: Bucket's not done yet.  You can't instantiate RGWSalBucket (pure virtual), so we have to create a RGWRadosBucket at this point.

File: src/rgw/rgw_rest_s3.h (hunk 6 above)
Comment (dang) [reply src/rgw/rgw_rest_s3.h]: Like I said, RGWSalBucket is temporary, and will be RGWBucket.

File: src/rgw/rgw_bucket.cc (hunk 2 above)
Comment (dang) [reply src/rgw/rgw_bucket.cc]: RGWSalBucket is renamed RGWBucket in the Bucket PR, to be submitted soon.

File: src/rgw/rgw_sal.h [hunk 7]
```diff
@@ -53,26 +53,85 @@ class RGWUser {
...
+				 std::map<RGWObjCategory, RGWStorageStats>& stats,
+				 std::string *max_marker = nullptr,
+				 bool *syncstopped = nullptr) = 0;
+    virtual int sync_user_stats(RGWBucketInfo &info) = 0;
+    virtual int update_container_stats(void) = 0;
+
+    std::string get_name() const { return ent.bucket.name; }
+    std::string get_tenant() const { return ent.bucket.tenant; }
+    std::string get_marker() const { return ent.bucket.marker; }
+    std::string get_bucket_id() const { return ent.bucket.bucket_id; }
```
Comment (cbodley) [inline src/rgw/rgw_sal.h]: can return these strings as const references

File: src/rgw/rgw_sal.h [hunk 8]
```diff
@@ -53,26 +53,85 @@ class RGWUser {
...
+  RGWBucketList& operator=(const RGWBucketList&) = default;
+  ~RGWBucketList();
+
+  map<string, RGWSalBucket*>& get_buckets() { return buckets; }
+  bool is_truncated(void) { return truncated; }
+  void set_truncated(bool trunc) { truncated = trunc; }
+  void add(RGWSalBucket* bucket) {
+    buckets[bucket->ent.bucket.name] = bucket;
+  }
+  size_t count() { return buckets.size(); }
```
Comment (cbodley) [inline src/rgw/rgw_sal.h]: count() and is_truncated() can be const

**Pull request merged**